packages/hadron-document/src/object-generator.ts

```diff
@@ -367,34 +412,62 @@ export class ObjectGenerator {
 
   // Return a $getField expression that evaluates to the current value
   // of the document at `path`.
-  private static createGetFieldExpr(path: string[]): BSONObject {
+  private static createGetFieldExpr(path: SubfieldDescription[]): BSONObject {
     return path.reduce(
-      (input, key) => ({
-        $getField: {
-          field: { $literal: key },
-          input,
-        },
-      }),
+      (input, { key, isArrayIndex }) =>
+        isArrayIndex
+          ? {
+              $arrayElemAt: [input, +key],
+            }
+          : {
+              $getField: {
+                field: { $literal: key },
+                input,
+              },
+            },
```

Comment:
In retrospect, it feels a bit dumb to have assumed (or probably just not thought about) that `$getField` and `$setField` don’t work on arrays the way they do on objects.